contracts/examples/bonding-curve/src/lib.rs

```diff
@@ -0,0 +1,325 @@
+				self.issue_failure_event(caller, message.err_msg.as_slice());
+				if token.is_egld() && amount > 0 {
+					self.send().direct_egld(caller, &amount, &[]);
+				}
+			},
+		}
+	}
+
+	#[endpoint(mintToken)]
+	fn mint_token(&self, amount: Self::BigUint) -> SCResult<AsyncCall<Self::SendApi>> {
+		only_owner!(self, "only owner may call this function");
+
+		let token = self.issued_token().get().identifier;
+		if self.issued_token().is_empty() {
+			return Err(SCError::from(BoxedBytes::from_concat(&[
+				token.as_esdt_identifier(),
+				b" was not issued yet",
+			])));
+		}
+
+		require!(
+			self.supply_type().get() != SupplyType::Limited
+				|| self.max_supply().get() > self.supply().get(),
+			"Maximum supply limit reached!"
+		);
+
+		require!(
+			self.supply_type().get() != SupplyType::Limited
+				|| self.max_supply().get() >= self.supply().get() + amount.clone(),
+			"Minting will exceed the maximum supply limit!"
+		);
+
+		let caller = self.blockchain().get_caller();
+		self.mint_started_event(&caller, &amount);
+
+		Ok(ESDTSystemSmartContractProxy::new_proxy_obj(self.send())
+			.mint(&token, &amount)
+			.async_call()
+			.with_callback(self.callbacks().mint_callback(&caller, &amount)))
+	}
+
+	#[callback]
+	fn mint_callback(
+		&self,
+		caller: &Address,
+		amount: &Self::BigUint,
+		#[call_result] result: AsyncCallResult<()>,
+	) {
+		match result {
+			AsyncCallResult::Ok(()) => {
+				self.mint_success_event(caller);
+				self.supply().update(|supply| *supply += amount);
+				self.balance().update(|balance| *balance += amount);
+			},
+			AsyncCallResult::Err(message) => {
+				self.mint_failure_event(caller, message.err_msg.as_slice());
+			},
+		}
+	}
+
+	// endpoints
+
+	fn check_sell_requirements(
+		&self,
+		token: &TokenIdentifier,
+		amount: &Self::BigUint,
+	) -> SCResult<()> {
+		let issued_token = self.issued_token().get().identifier;
+		if &issued_token != token {
+			return Err(SCError::from(BoxedBytes::from_concat(&[
+				b"Only ",
+				token.as_esdt_identifier(),
+				b" tokens accepted",
+			])));
+		}
+		if self.balance().get() < amount.clone() {
+			return Err(SCError::from(BoxedBytes::from_concat(&[
+				b"Contract does not have enough ",
+				token.as_esdt_identifier(),
+				b". Please try again once more is minted.",
+			])));
+		}
+		require!(
+			amount > &Self::BigUint::zero(),
+			"Must pay more than 0 tokens!"
+		);
+		Ok(())
+	}
+
+	fn check_buy_requirements(
+		&self,
+		token: &TokenIdentifier,
+		amount: &Self::BigUint,
+	) -> SCResult<()> {
+		let exchanging_token = &self.exchanging_token().get().identifier;
+		if exchanging_token != token {
+			return Err(SCError::from(BoxedBytes::from_concat(&[
+				b"Only ",
+				exchanging_token.as_esdt_identifier(),
+				b" tokens accepted",
+			])));
+		}
+		require!(
+			amount > &Self::BigUint::zero(),
+			"Must pay more than 0 tokens!"
+		);
+		Ok(())
+	}
+	fn get_curve_arguments(self) -> CurveArguments<Self::BigUint> {
```

Comment:
If you find yourself calling this function often, consider storing the whole struct under a single key. Doing a single big storage read is cheaper than reading parts of it step by step.